### PR TITLE
[DMS-41] Develop Automated Tests for API Schema File Problems

### PR DESCRIPTION
**Pull request by CSR2017**

### CODE REVIEW

File: src/EdFi.DataManagementService.Api.Tests.Integration/APISchemaFileTests.cs [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+                (collection) =>
+                {
+                    collection.AddTransient((x) => _apiSchemaProvider);
+                }
+            );
+        };
+    }
+
+    [Test]
+    public async Task Should_throw_error_when_no_resourcename_element()
```
Comment (stephenfuqua) [inline src/EdFi.DataManagementService.Api.Tests.Integration/APISchemaFileTests.cs:53]: Is there something special about "no_resourcename"? Why should we throw an error here?

Perhaps better to say that it is _responding_ with an error, since it is not _throwing_ an exception.

File: src/EdFi.DataManagementService.Api.Tests.Integration/EdFi.DataManagementService.Api.Tests.Integration.csproj [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="coverlet.collector" Version="6.0.0" />
+    <PackageReference Include="FakeItEasy" Version="8.1.0" />
+    <PackageReference Include="FluentAssertions" Version="6.12.0" />
+    <PackageReference Include="Microsoft.AspNetCore.Mvc.Testing" Version="8.0.1" />
+    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.8.0" />
+    <PackageReference Include="NUnit" Version="3.14.0" />
+    <PackageReference Include="NUnit.Analyzers" Version="3.9.0" />
```
Comment (stephenfuqua) [inline src/EdFi.DataManagementService.Api.Tests.Integration/EdFi.DataManagementService.Api.Tests.Integration.csproj:19]: Can we use NUnit 4?

File: src/EdFi.DataManagementService.Api.Tests.Integration/APISchemaFileTests.cs (hunk 1 above)
Comment (CSR2017) [reply src/EdFi.DataManagementService.Api.Tests.Integration/APISchemaFileTests.cs:53]: @stephenfuqua 
We are addressing all errors originating from the "ResourceSchema.cs" file. In the "InvalidResourceSchemas.json" file, we have included resource schemas for various invalid use cases. 
For example, the "no_resourcename" schema does not contain a "resourceName" element. So, 
errors will be thrown from the "ResourceSchema.cs" file when attempting to read the "resourceName" from the schema for a particular resource.

File: src/EdFi.DataManagementService.Api.Tests.Integration/APISchemaFileTests.cs [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+
+    [Test]
+    public async Task Should_throw_error_when_no_isdescriptor_element()
+    {
+        // Arrange
+        await using var factory = new WebApplicationFactory<Program>().WithWebHostBuilder(_webHostBuilder);
+        using var client = factory.CreateClient();
+
+        // Act
+        var response = await client.GetAsync("/data/ed-fi/noIsDescriptors");
```
[the remark this answers is not in the document]
Comment (CSR2017) [reply src/EdFi.DataManagementService.Api.Tests.Integration/APISchemaFileTests.cs:91]: @stephenfuqua 
On the frontend layer, there is no observable difference. The "CoreLoggingMiddleware" is to catch specific errors and consistently throw an "Internal server error." However, 
the specific error will be logged. Through these various tests, we are ensuring that invalid resource schemas are appropriately captured and throws error.

File: src/EdFi.DataManagementService.Api.Tests.Integration/APISchemaFileTests.cs (hunk 1 above)
Comment (stephenfuqua) [reply src/EdFi.DataManagementService.Api.Tests.Integration/APISchemaFileTests.cs]: This is helpful, so let's just add a comment in the class describing the scenarios so that we don't have to guess at it a year from now or when someone else needs to come in and edit it. Could use a block comment that describes the different scenarios in plain English as you've done above.

File: src/EdFi.DataManagementService.Api.Tests.Integration/APISchemaFileTests.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (stephenfuqua) [reply src/EdFi.DataManagementService.Api.Tests.Integration/APISchemaFileTests.cs]: I get the idea now that there are multiple problems in the bad schema file. I couldn't tell that from just reading the class before.

File: src/services/EdFi.DataManagementService.Api/Core/ApiSchema/ResourceSchema.cs [hunk 4]
```diff
@@ -303,13 +303,13 @@ public DocumentIdentity ExtractDocumentIdentity(JsonNode documentBody)
...
             ?? throw new InvalidOperationException(
-                "Expected subclassIdentityDocumentKey to be in documentPathsMapping, invalid ApiSchema"
+                "Expected subclassIdentityDocumentKey to be on ResourceSchema, invalid ApiSchema"
             );
 
         string superclassIdentityDocumentKey =
             _resourceSchemaNode["superclassIdentityDocumentKey"]?.GetValue<string>()
             ?? throw new InvalidOperationException(
-                "Expected superclassIdentityDocumentKey to be in documentPathsMapping, invalid ApiSchema"
+                "Expected superclassIdentityDocumentKey to be on ResourceSchema, invalid ApiSchema"
```
[the remark this answers is not in the document]
Comment (CSR2017) [reply src/services/EdFi.DataManagementService.Api/Core/ApiSchema/ResourceSchema.cs:312]: @bradbanister 
Quick question: Can I use "SelectRequiredNodeFromPathAs" extension method for retrieving value for specific element ? Should we consider adding the "Resource schema" name on the error message for making it easier to identify the invalid resource schema ?

File: src/EdFi.DataManagementService.Api.Tests.Integration/APISchemaFileTests.cs [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+
+[TestFixture]
+public class APISchemaFileTests
+{
+    /// <summary>
+    /// APISchemaFileTests contain tests to cover all errors that may arise from the "ResourceSchema.cs" file due to invalid resource schemas.
+    /// Within the "InvalidResourceSchemas.json" file, we have curated a collection of resource schemas representing various invalid use cases.
+    /// On the frontend layer, there is no observable difference. The "CoreLoggingMiddleware" is to catch specific errors and consistently throw an
+    /// "Internal server error." However, the specific error will be logged.Through these various tests,
+    /// we are ensuring that invalid resource schemas are appropriately captured and throws error.
```
Comment (stephenfuqua) [inline src/EdFi.DataManagementService.Api.Tests.Integration/APISchemaFileTests.cs:28]: 👍

**Review by stephenfuqua — APPROVED** [review]
